src/wheel/wheelfile.py

```diff
@@ -1,169 +1,311 @@
-        if mode == 'r' and not ef_name.endswith('/') and ef_name not in self._file_hashes:
-            raise WheelError("No hash found for file '{}'".format(native(ef_name)))
-
-        ef = ZipFile.open(self, name_or_info, mode, pwd)
-        if mode == 'r' and not ef_name.endswith('/'):
-            algorithm, expected_hash = self._file_hashes[ef_name]
-            if expected_hash is not None:
-                # Monkey patch the _update_crc method to also check for the hash from RECORD
-                running_hash = hashlib.new(algorithm)
-                update_crc_orig, ef._update_crc = ef._update_crc, _update_crc
-
-        return ef
-
-    def write_files(self, base_dir):
-        logger.info("creating '%s' and adding '%s' to it", self.filename, base_dir)
+                self._zip.close()
+            finally:
+                self._record_entries.clear()
+
+    def __enter__(self) -> 'WheelFile':
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.close()
+
+    @staticmethod
+    def _get_zipinfo_datetime(timestamp: float) -> Tuple[int, int, int, int, int, int]:
+        # Some applications need reproducible .whl files, but they can't do this without forcing
+        # the timestamp of the individual ZipInfo objects. See issue #143.
+        timestamp = int(os.environ.get('SOURCE_DATE_EPOCH', timestamp))
+        return time.gmtime(timestamp)[0:6]
+
+    def write_file(self, archive_name: str, contents: Union[bytes, str, PathLike]) -> None:
+        timestamp = time.time()
+        if isinstance(contents, bytes):
+            data = contents
+        elif isinstance(contents, str):
+            data = contents.encode('utf-8')
+        elif isinstance(contents, PathLike):
+            path = Path(contents)
+            timestamp = path.stat().st_mtime
+            data = path.read_bytes()
+        else:
+            raise TypeError('contents must be a str, bytes or a path-like object')
+
+        if archive_name not in self._exclude_archive_names:
+            hash_digest = hashlib.new(self._default_hash_algorithm, data).digest()
+            self._record_entries[archive_name] = WheelRecordEntry(
+                self._default_hash_algorithm, hash_digest, len(data))
+
+        zinfo = ZipInfo(archive_name, date_time=self._get_zipinfo_datetime(timestamp))
+        zinfo.external_attr = 0o664 << 16
+        self._zip.writestr(zinfo, data)
+
+    def write_data_file(self, archive_name: str, contents: Union[bytes, str, PathLike]) -> None:
+        archive_path = self._data_path + '/' + archive_name
+        self.write_file(archive_path, contents)
+
+    def write_metadata_file(self, archive_name: str,
+                            contents: Union[bytes, str, PathLike]) -> None:
+        archive_path = self._dist_info_path + '/' + archive_name
+        self.write_file(archive_path, contents)
+
+    def write_files_from_directory(self, base_path: Union[str, PathLike]) -> None:
```

Comment:
I'd suggest against adding this, in favor of minimizing the initial API surface.

I don't think having helper functions like this, "out-of-the-box" is necessary, and can be added afterwards. This can totally live independently as a `populate_wheel_from_directory` function that uses `obj.write_file`.

Reply:
Alright, I suppose I could remove it.

Reply:
I'm ambivalent. For my use cases I wouldn't use it, but I can see that build backends might find this convenient.

My suggestion would be to remove it from the initial version of the API, but it can be added later if it's found to be useful.

Reply:
Yea, I'm not saying this is never useful, but that we should start with a smaller API surface. We can then expand it based on feedback from users, but removing things is more difficult.

Reply:
I've removed it for now.

Reply:
Turns out this was used in at least 3 places (`bdist_wheel`, `wheel convert` and `wheel pack`).
Not having this is a bit inconvenient and I need to add _something_ in its place.